api/app/src/external/commonwell/patient.ts

```diff
@@ -0,0 +1,273 @@
+    );
+    throw new Error(msg);
+  }
+
+  // FIND OR CREATE PERSON
+  let commonwellPersonId;
+  try {
+    const person = makePersonForPatient(cwPatient);
+    commonwellPersonId = await findOrCreatePerson({ commonWell, queryMeta, person });
+  } catch (err) {
+    log(`Patient created @ CW but could not find/create Person`);
+    throw err;
+  }
+
+  // LINK THEM
+  await commonWell.patientLink(queryMeta, commonwellPersonId, patientRefLink);
+
+  // NETWORK LINKS
+  // TODO
+
+  return { commonwellPatientId, commonwellPersonId };
+}
+
+function makePersonForPatient(cwPatient: CommonwellPatient): CommonwellPerson {
+  return {
+    details: cwPatient.details,
+  };
+}
+
+// TODO finish alternative flows
+async function findOrCreatePerson({
+  commonWell,
+  queryMeta,
+  person,
+}: {
+  commonWell: CommonWell;
+  queryMeta: RequestMetadata;
+  person: CommonwellPerson;
+}): Promise<string> {
+  // SEARCH PERSON WITH STRONG ID OR DEMOGRAPHICS
+  // TODO #369
+  // TODO #369
+  // TODO #369
+
+  // IF NOT FOUND - ENROLL PERSON
+  console.log(`Enrolling this person: ${JSON.stringify(person, null, 2)}`); // TODO #369 remove this
+
+  const respPerson = await commonWell.enrollPerson(queryMeta, person);
+  console.log(`[CW REST PERSON ENROLL] `, respPerson); // TODO #369 remove this
+  const personId = getId(respPerson);
+  if (!personId) {
+    const msg = `Could not get person ID from CW response`;
+    console.log(`${msg} - CW response: ${JSON.stringify(respPerson)}`);
+    throw new Error(msg);
+  }
+  return personId;
+}
+
+export async function update(patient: Patient, facilityId: string): Promise<void> {
+  const log = Util.log(`Update CW Patient - ${patient.id}`);
+
+  const commonwellData = patient.data.externalData
+    ? (patient.data.externalData[ExternalMedicalPartners.COMMONWELL] as PatientDataCommonwell) // TODO validate the type
+    : undefined;
+  if (!commonwellData) throw new Error(`Missing commonwell data on patient ${patient.id}`);
+  const commonwellPatientId = commonwellData.patientId;
+  const commonwellPersonId = commonwellData.personId;
+
+  const { organization, facility } = await getPatientData(patient, facilityId);
+  const orgName = organization.data.name;
+  const orgId = organization.id;
+  const facilityNPI = facility.data["npi"] as string; // TODO #369 move to strong type - remove `as string`
+
+  const queryMeta = organizationQueryMeta(orgName, { npi: facilityNPI });
+  const cwPatient = patientToCommonwell({ patient, orgName, orgId });
+  const commonWell = makeCommonWellAPI(orgName, oid(orgId));
+
+  // UDPATE PATIENT
+  const respPatient = await commonWell.updatePatient(queryMeta, cwPatient, commonwellPatientId);
+  log(`[CW RESP PATIENT UPDATE] `, respPatient); // TODO #369 remove this
+
+  // UDPATE PERSON
+  try {
+    const person = makePersonForPatient(cwPatient);
+    const respPerson = await commonWell.updatePerson(queryMeta, person, commonwellPersonId);
```

Comment:
TODO to verify that there is still a person link to this patient here? I think `// REVIEW PERSON-PATIENT LINK` should be done before we try and update the person.

Reply:
Why do you think we should review the link before updating the person and patient? I think it wouldn't make any difference since we'll force the link to be LOLA2+ anyways, no?

Reply:
If a 3rd party removed the person from CW that used to exist and was linked to this patient, then there will be no person to update here.

Reply:
Ok, this is real problem, but its different from the original message:
1. Check link between person and patient, make sure its LOLA2+;
2. Recreate the person if it fails to update because of a 404.

### re: (1)

I can test to make sure, but we're storing the person ID, I'm assuming its the same for a given person regardless of the links it has with patients.

IFF that's the case, then this is safe.

Otherwise, I think this might be challenging for us to work around it... imagine:
- we create a new patient and person (essentially one entity @ metriport);
- someone updates CW removing the link between those two;
- our customer updates their patient;
- we query for the patient's person, and we find more than one.

If we can store the actual IDs of the entities @ CW, _in theory_ we would minimize any conflicts, mismatches, and OPs manually reviewing similar people.

### re: (2)

I think the way to go for this one would be to have error handling code that creates a new person if the update fails because of a 404, and update the person ID on our DB with the new one.

Thoughts about all of this?

Reply:
1 - I dont believe the person id's are different so i think we're good here

2 - Yea i think thats the best approach. I was thinking we could prompt the user if there is a 404 to select a person again but that wont really work via SDK

Reply:
Added back

Reply:
This makes sense - if we address (2) using that approach on 404, reviewing the person-patient link to make sure it's LOLA 2 afterwords works in all cases.

With this in mind, it's currently not clear to me what happens when someone "unenrolls" a person (delete is pretty obvious). Thinking about this scenario:
- We go to update a patient
- The underlying person is unenrolled, but still exists
- (Question): does the person still get updated, or do we have to re-enroll the person somehow beforehand? As per [CW docs](https://specification.commonwellalliance.org/services/patient-identity-and-linking/protocol-operations#871-person-enrollment): "The Person may still appear in searches but with its enrolled status set to False."
- After the update and/or re-enrollement, we'll also need to deal with the fact that the existing link is LOLA 1 and upgrade to LOLA 2.

Reply:
I just tested with person `6fbb3272-4ecc-4411-8332-368b3628881f`:
1. PUT `person/:id/unenroll` unenrolls the person
2. GET `person/:id` works (includes prop `enrolled: false`)
3. GET `person/:id/patientLink` returns 404
4. GET `/patient/:id/networkLink` returns 412 `Patient with Local Id ... is not linked to any Person`
5. PUT `person/:id/enroll` re-enrolls the person
6. (3) and (4) now return successfully